### PR TITLE
Feature/phpstan

### PR DESCRIPTION
**Pull request by schneiel**
install phpstan, fix warnings of phpstan, fix codestyle

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -27,12 +27,20 @@
...
     "require": {
         "typo3/cms-core": "~8.7 || ~9.5",
         "digedag/rn-base": ">=1.10.0"
     },
     "require-dev": {
-        "friendsofphp/php-cs-fixer": "~2@stable"
+        "friendsofphp/php-cs-fixer": "~2@stable",
+        "phpstan/phpstan": "^0.11",
+        "saschaegerer/phpstan-typo3": "dev-feature/update-phpstan",
+        "typo3/testing-framework": "*"
```
Comment (Baachi) [inline composer.json]: Please use a specific version and not the "star of death" :smile:

File: composer.json [hunk 2]
```diff
@@ -27,12 +27,20 @@
         "mksanitizedparameters": "self.version",
         "typo3-ter/mksanitizedparameters": "self.version"
     },
+    "repositories": [{
```
Comment (Baachi) [inline composer.json:30]: This should be resolved before we merge that PR

File: hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php [hunk 3]
```diff
@@ -63,7 +63,10 @@ public function sanitizeGlobalInputArrays()
      */
     protected function getMksanitizedparametersMainClass()
     {
-        return tx_rnbase::makeInstance('tx_mksanitizedparameters');
+        /** @var tx_mksanitizedparameters $class */
+        $class = tx_rnbase::makeInstance('tx_mksanitizedparameters');
```
Comment (Baachi) [inline hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php:67]: Why is this needed? This is bad practice in my opinion. Use a variable just for a `@var` statement is not worth it.
Can you check if our package [`phpstan-rnbase`](https://github.com/DMKEBUSINESSGMBH/phpstan-rnbase/) fix this problem?

File: util/class.tx_mksanitizedparameters_util_RegularExpression.php [hunk 4]
```diff
@@ -30,8 +30,6 @@ class tx_mksanitizedparameters_util_RegularExpression
     /**
      * @param string $pattern
      * @param mixed  $value
-     *
-     * @return string
```
Comment (Baachi) [inline util/class.tx_mksanitizedparameters_util_RegularExpression.php:34]: Wouldn't be better to specify the correct return type, rather to delete them?

**Review by Baachi — CHANGES_REQUESTED** [review]

File: composer.json (hunk 2 above)
Comment (schneiel) [reply composer.json:30]: Of course. Incidentally Sascha Egerer commented your PR https://github.com/sascha-egerer/phpstan-typo3/pull/10.

File: composer.json (hunk 1 above)
Comment (schneiel) [reply composer.json]: You're right. I've tried to find a suitable version of typo3/testing-framework, but version 2.0.4, which support typo3/cms-core:"^8.5 || ^9", not support phpunit 7. I think, downgrade phpunit isn't the way to go.

File: hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php (hunk 3 above)
Comment (schneiel) [reply hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php:67]: Your package don't fix this problem. I can't think of a better solution for this problem.

Can you add this fix to your package, please?

File: util/class.tx_mksanitizedparameters_util_RegularExpression.php (hunk 4 above)
Comment (schneiel) [reply util/class.tx_mksanitizedparameters_util_RegularExpression.php:34]: You're right again. This method is called in **tx_mksanitizedparameters_sanitizer_Alnum** and **tx_mksanitizedparameters_sanitizer_Alpha**. Those methods only called in TestCases.

Can I change the return type of those methodes to null|string|string[] ?

But I think, this is also the wrong way, because those methods cast $value to string. Only one string can be returned like this.

File: composer.json (hunk 2 above)
Comment (Baachi) [reply composer.json:30]: Yeah, but this PR should be merged first. Using a fork stinks :P 
I will try push my PR forwards.

File: hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php (hunk 3 above)
Comment (Baachi) [reply hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php:67]: Can you open an issue for this, with a reproduceable testcase?

File: class.tx_mksanitizedparameters.php [hunk 5]
```diff
@@ -354,8 +354,9 @@ private function injectCommonRulesFromCurrentIntoNextLevelIfNotSet(
     }
 
     /**
-     * @param array $rules
-     * @param mixed $defaultRules
+     * @param array $rulesForValue
+     * @param mixed $allRules
```
Comment (hannesbochmann) [inline class.tx_mksanitizedparameters.php]: You could force $allRules to be an array by adding the appropriate type hint in the parameter signature of the method. It can never be something else if the code of the method should work as the variable is accessed as an array.

File: util/class.tx_mksanitizedparameters_util_RegularExpression.php (hunk 4 above)
Comment (hannesbochmann) [reply util/class.tx_mksanitizedparameters_util_RegularExpression.php:34]: I think null|string|array would be correct as that's what preg_replace returns as stated in the PHP documentation (https://php.net/manual/de/function.preg-replace.php).

**Review by hannesbochmann — CHANGES_REQUESTED** [review]

File: hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php (hunk 3 above)
Comment (schneiel) [reply hooks/class.tx_mksanitizedparameters_hooks_PreprocessTypo3Requests.php:67]: [https://github.com/DMKEBUSINESSGMBH/phpstan-rnbase/issues/2](https://github.com/DMKEBUSINESSGMBH/phpstan-rnbase/issues/2)

DONE

File: util/class.tx_mksanitizedparameters_util_RegularExpression.php (hunk 4 above)
Comment (schneiel) [reply util/class.tx_mksanitizedparameters_util_RegularExpression.php:34]: @hbochmann 
You're right. I will do that!

Comment (schneiel) [reply util/class.tx_mksanitizedparameters_util_RegularExpression.php:34]: @hbochmann 

But please bear in mind, that if change the typhint of **callPregReplace**, I also must change the typhints of those methods, which call this.

Those methods deliver a value, which is parsed to string. So those methods only can return a string (see the documentation).

There I can use a annotation, but parsing is not possible.

File: class.tx_mksanitizedparameters.php (hunk 5 above)
Comment (schneiel) [reply class.tx_mksanitizedparameters.php]: You're right. That i will do!
